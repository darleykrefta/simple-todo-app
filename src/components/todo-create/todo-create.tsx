import React, { useCallback } from 'react'

import { useRecoilValue, useSetRecoilState } from 'recoil'

import { Badge, Button, Flex, Input, Text } from '@chakra-ui/react'

import { generateId } from 'helpers/functions'
import { todoItemsQuantityStore, todoItemsStore } from 'stores/todo-store'

export const TodoCreate: React.FC = () => {
  const [value, setValue] = React.useState<string>('')

  const setTodoItems = useSetRecoilState(todoItemsStore)
  const { totalCompletedItems, totalUncompletedItems } = useRecoilValue(todoItemsQuantityStore)

  const handleAddItem = useCallback(() => {
    if (!value) {
      return
    }

    setTodoItems(oldTodoItems => [
      ...oldTodoItems,
      {
        id: generateId(),
        description: value,
        completed: false
      }
    ]),
      setValue('')
  }, [value, setTodoItems])

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => setValue(event.target.value)

  return (
    <Flex flexDir="column" alignItems="center">
      <Text color="primary" fontSize="36px" py={5}>
        TODO
      </Text>

      <Flex w="320px" justifyContent="space-around" my={5}>
        <Badge fontSize="sm" background="primary" color="white">
          <Text textDecoration="line-through" data-testid="completed-label">
            completed {totalCompletedItems}
          </Text>
        </Badge>

        <Badge fontSize="sm" background="primary" color="white" data-testid="uncompleted-label">
          uncompleted {totalUncompletedItems}
        </Badge>
      </Flex>

      <Flex w="320px" justifyContent="space-around">
        <Input
          w="220px"
          borderColor="primary"
          background="white"
          size="lg"
          value={value}
          onChange={handleChange}
          data-testid="input-todo-add"
        />

        <Button
          background="primary"
          color="white"
          variant="solid"
          size="lg"
          onClick={() => handleAddItem()}
          data-testid="button-todo-add"
        >
          ADD
        </Button>
      </Flex>
    </Flex>
  )
}
