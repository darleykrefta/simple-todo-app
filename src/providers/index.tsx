import React from 'react'

import { RecoilRoot } from 'recoil'

import { ChakraProvider, CSSReset, extendTheme } from '@chakra-ui/react'

const theme = extendTheme({
  colors: {
    primary: '#590202',
    secondary: '#F26D3D',
    background: '#84BF9E'
  }
})

export const Providers: React.FC = ({ children }) => {
  return (
    <ChakraProvider theme={theme}>
      <RecoilRoot>
        <CSSReset />
        {children}
      </RecoilRoot>
    </ChakraProvider>
  )
}
