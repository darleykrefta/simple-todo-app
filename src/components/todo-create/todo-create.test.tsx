import React from 'react'

import { cleanup, fireEvent, render, screen } from '@testing-library/react'

import { Providers } from 'providers'
import { TodoCreate } from 'components/todo-create'

const setup = () => render(<TodoCreate />, { wrapper: Providers })

describe('TodoCreate component', () => {
  afterEach(cleanup)

  it('should be valid with last version', () => {
    const { asFragment } = setup()
    expect(asFragment()).toMatchSnapshot()
  })

  it('should be initiate with zero completed values', () => {
    setup()
    const completedLabel = screen.getByTestId('completed-label')
    expect(completedLabel).toHaveTextContent('completed 0')
  })

  it('should be initiate with zero uncompleted values', () => {
    setup()
    const uncompletedLabel = screen.getByTestId('uncompleted-label')
    expect(uncompletedLabel).toHaveTextContent('uncompleted 0')
  })

  it('should validate add event within value', () => {
    setup()
    const buttonAdd = screen.getByTestId('button-todo-add')
    fireEvent.click(buttonAdd)
    const uncompletedLabel = screen.getByTestId('uncompleted-label')
    expect(uncompletedLabel).toHaveTextContent('uncompleted 0')
  })

  it('should validate input change', () => {
    setup()
    const value = 'something'
    const inputAdd = screen.getByTestId('input-todo-add')
    const buttonAdd = screen.getByTestId('button-todo-add')
    const uncompletedLabel = screen.getByTestId('uncompleted-label')
    fireEvent.change(inputAdd, { target: { value } })
    fireEvent.click(buttonAdd)
    expect(uncompletedLabel).toHaveTextContent('uncompleted 1')
  })
})
