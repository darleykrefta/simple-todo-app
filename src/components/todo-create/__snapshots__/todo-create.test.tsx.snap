// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoCreate component should be valid with last version 1`] = `
<DocumentFragment>
  <div
    class="css-1vr2anf"
  >
    <p
      class="chakra-text css-1hdr793"
    >
      TODO
    </p>
    <div
      class="css-uvr4ox"
    >
      <span
        class="chakra-badge css-1lzl5c4"
      >
        <p
          class="chakra-text css-4jky3p"
          data-testid="completed-label"
        >
          completed 0
        </p>
      </span>
      <span
        class="chakra-badge css-1lzl5c4"
        data-testid="uncompleted-label"
      >
        uncompleted 0
      </span>
    </div>
    <div
      class="css-n6xhs7"
    >
      <input
        class="chakra-input css-kp1c15"
        data-testid="input-todo-add"
        value=""
      />
      <button
        class="chakra-button css-xfuke2"
        data-testid="button-todo-add"
        type="button"
      >
        ADD
      </button>
    </div>
  </div>
</DocumentFragment>
`;
